esp-hal/src/gpio/mod.rs

```diff
@@ -961,80 +967,64 @@ where
     }
 }
 
-/// General Purpose Input/Output driver
-pub struct Io {
-    _io_mux: IO_MUX,
-    /// The pins available on this chip
-    pub pins: Pins,
-}
-
-impl Io {
-    /// Initialize the I/O driver.
-    pub fn new(gpio: GPIO, io_mux: IO_MUX) -> Self {
-        Self::new_with_priority(gpio, io_mux, crate::interrupt::Priority::min())
-    }
-
-    /// Initialize the I/O driver with a interrupt priority.
-    ///
-    /// This decides the priority for the interrupt when using async.
-    pub fn new_with_priority(
-        mut gpio: GPIO,
-        io_mux: IO_MUX,
-        prio: crate::interrupt::Priority,
-    ) -> Self {
-        gpio.bind_gpio_interrupt(gpio_interrupt_handler);
-        crate::interrupt::enable(crate::peripherals::Interrupt::GPIO, prio).unwrap();
-
-        Self::new_no_bind_interrupt(gpio, io_mux)
-    }
-
-    /// Initialize the I/O driver without enabling the GPIO interrupt or
-    /// binding an interrupt handler to it.
-    ///
-    /// *Note:* You probably don't want to use this, it is intended to be used
-    /// in very specific use cases. Async GPIO functionality will not work
-    /// when instantiating `Io` using this constructor.
-    pub fn new_no_bind_interrupt(gpio: GPIO, _io_mux: IO_MUX) -> Self {
-        Io {
-            _io_mux,
-            pins: gpio.pins(),
+impl<const GPIONUM: u8> InterruptConfigurable for GpioPin<GPIONUM>
+where
+    Self: GpioProperties,
+{
+    fn set_interrupt_handler(&mut self, handler: InterruptHandler) {
+        raise_gpio_interrupt_priority(handler.priority());
```

Comment:
Hmm, I have a vauge feeling this won't play nicely with this idea.  https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/intr_alloc.html#iram-safe-interrupt-handlers .

There should be some way to force users to choose a single priority from the start.

Reply:
I'm slightly afraid "the priority you set here will be ignored" isn't what we are aiming for, but it's something to consider.

Reply:
Or perhaps this shouldn't take `InterruptHandler` as a parameter? Since this isn't a real interrupt and that way there's nothing to ignore here